cnd/src/db/tables.rs

```diff
@@ -0,0 +1,733 @@
+
+            let swap: Swap = swaps::table
+                .filter(swaps::local_swap_id.eq(key))
+                .first(connection)?;
+
+            let insertable = ISecretHash {
+                swap_id: swap.id,
+                secret_hash: Text(secret_hash),
+            };
+
+            diesel::insert_into(secret_hashes::dsl::secret_hashes)
+                .values(insertable)
+                .execute(&*connection)
+        })
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn load_secret_hash(
+        &self,
+        swap_id: LocalSwapId,
+    ) -> anyhow::Result<rfc003::SecretHash> {
+        let record: SecretHash = self
+            .do_in_transaction(|connection| {
+                let key = Text(swap_id);
+
+                let swap: Swap = swaps::table
+                    .filter(swaps::local_swap_id.eq(key))
+                    .first(connection)?;
+
+                SecretHash::belonging_to(&swap).first(connection).optional()
+            })
+            .await?
+            .ok_or(Error::SwapNotFound)?;
+
+        Ok(record.secret_hash.0)
+    }
+
+    pub async fn save_shared_swap_id(
+        &self,
+        swap_id: LocalSwapId,
+        shared_swap_id: swap_protocols::SharedSwapId,
+    ) -> anyhow::Result<()> {
+        self.do_in_transaction(|connection| {
+            let key = Text(swap_id);
+
+            let swap: Swap = swaps::table
+                .filter(swaps::local_swap_id.eq(key))
+                .first(connection)?;
+
+            let insertable = ISharedSwapId {
+                swap_id: swap.id,
+                shared_swap_id: Text(shared_swap_id),
+            };
+
+            diesel::insert_into(shared_swap_ids::dsl::shared_swap_ids)
+                .values(insertable)
+                .execute(&*connection)
+        })
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn load_shared_swap_id(
+        &self,
+        swap_id: LocalSwapId,
+    ) -> anyhow::Result<swap_protocols::SharedSwapId> {
+        let record: SharedSwapId = self
+            .do_in_transaction(|connection| {
+                let key = Text(swap_id);
+
+                let swap: Swap = swaps::table
+                    .filter(swaps::local_swap_id.eq(key))
+                    .first(connection)?;
+
+                SharedSwapId::belonging_to(&swap)
+                    .first(connection)
+                    .optional()
+            })
+            .await?
+            .ok_or(Error::SwapNotFound)?;
+
+        Ok(record.shared_swap_id.0)
+    }
+
+    pub async fn save_address_hint(
+        &self,
+        swap_id: LocalSwapId,
+        address_hint: libp2p::Multiaddr,
+    ) -> anyhow::Result<()> {
+        self.do_in_transaction(|connection| {
+            let key = Text(swap_id);
+
+            let swap: Swap = swaps::table
+                .filter(swaps::local_swap_id.eq(key))
+                .first(connection)?;
+
+            // Needed because Multiaddr does not implement Copy
```

Comment:
Clone it then? :)

Reply:
that did not work, but I can try again.

Reply:
Where does this bound come from? Should be in the error message somewhere.

Reply:
Used 
```
            let insertable = IAddressHint {
                swap_id: swap.id,
                address_hint: Text(address_hint.clone()),
            };
```